### ISSUE TITLE
How to Run Multiple Rest Scripts at Once

### ISSUE THREAD

**Issue by kool2zero** [cubewise-code/TM1py#42]
Is there a specific way to get TM1py to run in concurrent sessions when connecting through TM1Service?
We created a script that reads a CSV file and kicks off multiple TIs via rest.  However, we want to kick off a second instance of this script with different TIs at the same time but it seems that it is not working.  We tried looking at session_id but are unsure of how to generate a unique one.

Any help would be great!

Code attached.
[RunTI.txt](https://github.com/cubewise-code/TM1py/files/2119761/RunTI.txt)

**Comment by MariusWirtz** [comment]
Hi,

there is no "specific" way to do it. It should work like in this sample:
https://github.com/cubewise-code/TM1py-samples/blob/master/Other/run%20TI%20Processes%20in%20parallel.py

What is the error that you get when you run the below code?

I don't think the session_id is causing the issue.
You don't need to create a session_id yourself. When you instantiate the TM1Service it authenticates and opens a TM1-session. Then it re-uses this session for every subsequent call.
You can do calls concurrently with only one session_id.

**Comment by kool2zero** [comment]
The error I get is 401 Unauthorized.
When I call my RunTI.py script by itself it works fine.  When it gets called simultaneously with different file lists passed it, one of the executions of the script runs fine, the other gets the 401 error.

**Comment by MariusWirtz** [comment]
Hi,

TM1 gives you the 401 Unauthorized Status, when you
a) pass wrong authentication credentials or 
b) when the Session is timed-out or closed

I think you might need to look into b). 
When you instantiate the TM1Service through the `with` statement it will close the session autonomously once the "block" is complete.
It kinda works the same as when you read a file through `with` in python...

Do the two scripts perhaps share the same session? Perhaps when the first script is complete it closes the session and the second script then gets an Unauthorized error...

If you don't want the TM1py to close the session for you, you can instantiate the TM1Service like this:
`tm1 = TM1Service(address="10.77.19.60", port=12354, user="admin", password="apple", ssl=True)`

and do the logout manually when the session is no longer needed:
`tm1.logout()`

**Comment by kool2zero** [comment]
Would calling the tm1 service in both executions of the same script return two different sessions?  So if I did the explict call like you suggest above, would calling tm1.logout in one of the executions log out the other?

**Comment by MariusWirtz** [comment]
No. One logout should not logout the other session.

**Comment by kool2zero** [comment]
It looks like replacing the `with` with `tm1 = TM1Service(address="10.77.19.60", port=12354, user="admin", password="apple", ssl=True)` fixed it.  Thanks for your help!

**Comment by MariusWirtz** [comment]
I created an open source project on Github to replace Hustle / RunTI with a REST API based solution:
https://github.com/MariusWirtz/RushTI
It's in an early stage, but maybe this is interesting for you. Feel free to contribute or open issues
